Islandora/islandora_scholar
# ISLANDORA-2172: Take advantage of content negociation capabilities

**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2172

# What does this Pull Request do?

Currently, [islandora_doi](https://github.com/Islandora/islandora_scholar/tree/7.x/modules/doi) uses [Crossref](https://www.crossref.org)'s openurl to retrieve metadata. This requires users to set up an account with Crossref.

Instead, islandora_doi can take advantage of [content negociation](https://www.crosscite.org/docs.html) and retrieve the metadata directly by querying https://doi.org. Not only does this method work without prior account creation, it also has the benefit of working with other metadata providers (e.g. [DataCite](https://datacite.org), [mEDRA](https://www.medra.org), etc...). This makes islandora_doi fit for future extensions.

This pull request adds this possibility, along with improving error messages.

# What's new?

* Introduce a new mode for retrieving metadata featuring:
   * Use content negociation rather than openurl
   * No prior account creation required
   * Prepare for extending to metadata providers other than Crossref
      (e.g. DataCite, mEDRA)
* Streamline error handling:
   * Handle errors exclusively in modules/doi/include/utilities.inc in order to
      obtain same messages regardless of ingest method (populator/importer)
   * Especially in the new non-legacy mode, failures to ingest should now give
      more useful messages (wrong type, wrong metadata provider, etc...)

The changes should be pretty non-intrusive. In particular, all existing
functions have been preserved (no API change for those who created their own
extensions). Also, users who already have their account set up with Crossref
will default to keep querying in legacy mode. Only the improvements in error
handling should be visible. New users will default to querying in the new mode.
The querying behaviour can be changed at all times on the configuration page
/admin/islandora/solution_pack_config/scholar/islandora_doi.


# How should this be tested?

As always, never test in a production environment!

Steps:
* Test importer & populator on different DOIs (including DOIs of metadata providers others than Crossref, as well as other document types such as books)
* Merge this PR and re-test. Everything but the error messages should be the same (the latter being a bit more useful)
* Go to /admin/islandora/solution_pack_config/scholar/islandora_doi and de-select legacy mode
* Re-test and see better error messages
* Delete the variable 'islandora_doi_openurl_pid' (or set the PID to the default "user@example.com") and re-test: in non-legacy mode the metadata is still retrievable

# Additional Notes:

* Does this change require documentation to be updated? **no**
* Does this change add any new dependencies? **no**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? **no**
* Could this change impact execution of existing code? **no**

# Interested parties
@bryjbrown @DonRichards @Islandora/7-x-1-x-committers
